engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/IsValueOfTypeNode.java

```diff
@@ -0,0 +1,96 @@
+package org.enso.interpreter.node.expression.builtin.meta;
+
+import com.oracle.truffle.api.dsl.Fallback;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.nodes.NodeInfo;
+import com.oracle.truffle.api.profiles.ConditionProfile;
+import org.enso.interpreter.node.expression.builtin.meta.IsSameObjectNode;
+import org.enso.interpreter.node.expression.builtin.meta.TypeOfNode;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.callable.atom.AtomConstructor;
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
+import org.enso.interpreter.runtime.type.TypesGen;
+
+/** An implementation of the payload check against the expected panic type. */
+@NodeInfo(shortName = "IsValueOfTypeNode")
+public abstract class IsValueOfTypeNode extends Node {
+  private @Child IsSameObjectNode isSameObject = IsSameObjectNode.build();
+  private @Child TypeOfNode typeOfNode = TypeOfNode.build();
+  private final ConditionProfile profile = ConditionProfile.createCountingProfile();
+
+  public static IsValueOfTypeNode build() {
+    return IsValueOfTypeNodeGen.create();
+  }
+
+  public abstract boolean execute(Object panicType, Object payload);
+
+  @Specialization
+  boolean doAtom(AtomConstructor consr, Object payload) {
+    Object panicType = consr.getType();
+    return execute(panicType, payload);
+  }
+
+  @Specialization(guards = {"interop.isMetaObject(panicType)"})
```

Comment:
I am surprised to not see specialization for `Type`. There is [one in is_a PR](https://github.com/enso-org/enso/pull/3949/files#r1051858704).

Reply:
That reshuffle revealed also a bug in one of our tests when
```
case Array of
    Array_Module -> 1
    _ : Array -> 2
    Array -> 3
    _ -> 4
```
returned `2` instead of `3`. Due to the way we have to deal with arrays this wasn't a straightforward change but we now specialize on `Type`.